src/cli/rustup_mode.rs

```diff
@@ -76,11 +75,11 @@ pub fn main() -> Result<utils::ExitCode> {
     let matches = match cli().try_get_matches_from(process().args_os()) {
         Ok(matches) => Ok(matches),
         Err(err) if err.kind() == DisplayHelp => {
-            write!(process().stdout().lock(), "{err}")?;
```

Comment:
If I remember correctly the use of `.lock()` was needed to avoid race conditions, because stdout returns a thread-usable, but not locked, output; and write! is not atomic.

Reply:
Some parts of our code are locked and some parts are not. What is the basis for it?

Reply:
This is what happens with unlocked terminals and threads:
```
ColorSpec { fg_color: Some(Cyan), bg_color: None, bold: false, intense: false, underline: false, dimmed: false, italic: false, reset: true, strikethrough: false } text!
ColorSpec { fg_color: Some(Cyan), bg_color: None, bold: false, intense: false, underline: false, dimmed: false, italic: false, reset: true, strikethrough: false } text!
ColorSpec { fg_color: Some(Cyan), bg_color: None, bold: false, intense: false, underline: false, dimmed: false, italic: false, reset: true, strikethrough: false } text!
ColorSpec { fg_color: Some(Cyan), bg_color: None, bold: false, intense: ColorSpec { fg_color: Some(Green), bg_color: None, bold: false, intense: false, underline: false, dimmed: false, italic: false, reset: true, strikethrough: false } text!
ColorSpec { fg_color: Some(Green), bg_color: None, bold: false, intense: false, underline: false, dimmed: false, italic: false, reset: true, strikethrough: false } text!
ColorSpec { fg_color: Some(Green), bg_color: None, bold: false, intense: false, underline: false, dimmed: false, italic: false, reset: true, strikethrough: false } text!
ColorSpec { fg_color: Some(Green), bg_color: None, bold: false, intense: false, underline: false, dimmed: false, italic: false, reset: true, strikethrough: false } text!
ColorSpec { fg_color: Some(Green), bg_color: None, bold: false, intense: false, underline: false, dimmed: false, italic: false, reset: true, strikethrough: false } text!
```

Generated from https://github.com/rbtcollins/console-race

Reply:
If you `cargo run -- 1` instead of simple cargo run, you'll get a version using locked streams, and that will not mix up colors or lines at all. Example above showing the problem was generated on Linux.

Basically *everything* that is writing a non-atomic unit of IO to the console must use a locked writer.

So for instance a single refresh of a multiline progress bar must run against a locked stream.
The outputting of a status message that makes exactly one IO call - e.g. 'foo bar\n' - is implicitly locked in some cases - the write() syscall is generally defined as race-free - it either writes some bytes or fails, but doesn't write some bytes interleaved with another write() syscall. The problem is that write! and friends don't make a single syscall. They make many. And higher level abstractions like termcolor make even more syscalls. One for setting the color, one for the content being written.

So at the file descriptor level we get:
write(1, b"<color escape code");
write(1, b"Color Spec {");
write(1, b"fg_color: ");
write(1, b"Some(");
write(1, b"Cyan");
write(1, b"), ");

And when two threads are doing that at the same time, any one thread is sequential, but the shared file descriptor can see any of those from any thread with any arbitrary interleaving.

Reply:
Fixed in https://github.com/rust-lang/rustup/pull/3351/commits/8af7e6a4d0d85ab9c4c8af43db4b2af5d8868a2d

Reply:
Maybe we can improve it in the future. We could pass a context to each function that contains the terminal and provides the `info`, `error`, and `warn` functions. This way we don't need to go around accessing process().stdout().

Reply:
Maybe! The underlying challenge is this: rust std defines stdout / stderr / stdin as thread-agnostic globals (modelled via a function wrapping a OnceLock) rather than a File which would have the normal borrow checker constraints over it.

So there is an accessible-via-any thread struct with interior mutability as the *contract* for interacting with the terminal.

Then CurrentProcess comes in, which seeks to replace this global state with isolation, to permit concurrent tests without the overhead of forking/execing :- it has to meet the existing contract, because other libraries and code often want to work that way.